src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,382 @@
+            TMethodDeclarationSyntax,
+            TStatementSyntax,
+            TExpressionSyntax,
+            TInvocationSyntax>
+        : CodeRefactoringProvider
+        where TMethodDeclarationSyntax : SyntaxNode
+        where TStatementSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+        where TInvocationSyntax : TExpressionSyntax
+    {
+        /// <summary>
+        /// A preferred name used to generated a declaration when the
+        /// inline method's body is not a valid expresion in ExpressionStatement
+        /// Example:
+        /// void Caller()
+        /// {
+        ///     Callee();
+        /// }
+        /// int Callee()
+        /// {
+        ///     return 1;
+        /// };
+        /// After it should be:
+        /// void Caller()
+        /// {
+        ///     int temp = 1;
+        /// }
+        /// int Callee()
+        /// {
+        ///     return 1;
+        /// };
+        /// '1' is not a valid expression in ExpressionStatement so a declaration is needed to be generated.
+        /// </summary>
+        private const string TemporaryName = "temp";
+
+        private readonly ISyntaxFacts _syntaxFacts;
+        private readonly ISemanticFactsService _semanticFactsService;
+
+        protected abstract TExpressionSyntax? GetInlineExpression(TMethodDeclarationSyntax calleeMethodDeclarationSyntaxNode);
+        protected abstract SyntaxNode? GetEnclosingMethodLikeNode(SyntaxNode syntaxNode);
+        protected abstract SyntaxNode GenerateTypeSyntax(ITypeSymbol symbol, bool allowVar);
+
+        /// <summary>
+        /// Check if <paramref name="expressionNode"/> could be used as an Expression in ExpressionStatement
+        /// </summary>
+        protected abstract bool IsValidExpressionUnderStatementExpression(TExpressionSyntax expressionNode);
+
+        /// <summary>
+        /// Check if <paramref name="syntaxNode"/> could be replaced by ThrowExpression.
+        /// For VB it always return false because ThrowExpression doesn't exist.
+        /// </summary>
+        protected abstract bool CanBeReplacedByThrowExpression(SyntaxNode syntaxNode);
+
+        protected AbstractInlineMethodRefactoringProvider(
+            ISyntaxFacts syntaxFacts,
+            ISemanticFactsService semanticFactsService)
+        {
+            _syntaxFacts = syntaxFacts;
+            _semanticFactsService = semanticFactsService;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, _, cancellationToken) = context;
+            var calleeMethodInvocationNode = await context.TryGetRelevantNodeAsync<TInvocationSyntax>().ConfigureAwait(false);
+            if (calleeMethodInvocationNode == null)
+            {
+                return;
+            }
+            var statementContainsCallee = calleeMethodInvocationNode.GetAncestor<TStatementSyntax>();
+            if (statementContainsCallee == null)
```

Comment:
say you have:  `Action a = () => Foo();`

And we're inlining `Foo()`.  the statement node here will be the `Action a = ...` one.  Is that the right thing?

Reply:
Is 'Foo()' is an ExpressionStatement?

Reply:
Ah, checked the syntax tree and it is not. Then I think this will need to be treat differently.

Reply:
I see that if it is `Action a = () => Foo(Bar())` then it might need to become `Action a = () => { int t = Bar(); Return Foo(); }`